db/version_set.cc

```diff
@@ -3297,6 +3299,132 @@ Status VersionSet::LogAndApply(
+          edit, &have_log_number, &log_number, &have_prev_log_number,
+          &previous_log_number, &have_next_file, &next_file,
+          &have_last_sequence, &last_sequence, &min_log_number_to_keep,
+          &max_column_family);
+      if (!s.ok()) {
+        break;
+      }
+      if (column_family_set_->get_table_cache()->GetCapacity() ==
+          TableCache::kInfiniteCapacity) {
+        // Unlimited table cache. Pre-load table handle now so that the table
+        // files are still accessible to us after the primary unlinks them.
+        auto builder_iter = active_version_builders_.find(edit.column_family_);
+        assert(builder_iter != active_version_builders_.end());
+        auto builder = builder_iter->second->version_builder();
+        assert(builder != nullptr);
+        s = builder->LoadTableHandlers(
+            cfd->internal_stats(), db_options_->max_file_opening_threads,
+            false /* prefetch_index_and_filter_in_cache */,
+            false /* is_initial_load */,
+            cfd->GetLatestMutableCFOptions()->prefix_extractor.get());
+        if (!s.ok() && !s.IsPathNotFound()) {
+          break;
+        } else if (s.IsPathNotFound()) {
+          s = Status::OK();
+          // TODO (yanqin) release file descriptors already opened, or modify
+          // LoadTableHandlers so that opened files are not re-opened.
+        } else {  // s.ok() == true
+          auto version = new Version(cfd, this, env_options_,
+                                     *cfd->GetLatestMutableCFOptions(),
+                                     current_version_number_++);
+          builder->SaveTo(version->storage_info());
+          version->PrepareApply(*cfd->GetLatestMutableCFOptions(), true);
+          AppendVersion(cfd, version);
+          active_version_builders_.erase(builder_iter);
+          if (cfds_changed->count(cfd) == 0) {
+            cfds_changed->insert(cfd);
+          }
+        }
+      }
+      if (have_next_file) {
+        next_file_number_.store(next_file + 1);
+      }
+      if (have_last_sequence) {
+        last_allocated_sequence_ = last_sequence;
+        last_published_sequence_ = last_sequence;
+        last_sequence_ = last_sequence;
+      }
+      if (have_prev_log_number) {
+        prev_log_number_ = previous_log_number;
+        MarkFileNumberUsed(previous_log_number);
+      }
+      if (have_log_number) {
+        MarkFileNumberUsed(log_number);
+      }
+      column_family_set_->UpdateMaxColumnFamily(max_column_family);
+      MarkMinLogNumberToKeep2PC(min_log_number_to_keep);
+    }
+    if (s.ok()) {
+      // It's possible that we have finished reading the current MANIFEST, and
+      // the primary has created a new MANIFEST.
+      log::Reader::Reporter* reporter = reader->GetReporter();
+      s = MaybeSwitchManifest(reporter, manifest_reader);
+      reader = manifest_reader->get();
+    }
+    if (s.ok() && reader->file()->file_name() == old_manifest_path) {
```

Comment:
As long as the secondary has opened the manifest, it's ok if the primary deletes it because the secondary should still be able to read its content before closing it.